### ISSUE TITLE
Lack of `childViewContainer` in Marionette 3 makes valid table layouts difficult.

### ISSUE THREAD

**Issue by jgerigmeyer** [marionettejs/backbone.marionette#3150]
### Description

I have a table layout (`<table>`) where I would like each item in a `collection` to render as a `<tbody>`. For example, the goal is markup such as:

```html
<!-- for a collection with two models -->
<table>
  <thead>
    ...
  </thead>
  <!-- first model -->
  <tbody>
    <tr>
      ...
    </tr>
    <tr>
      ...
    </tr>
  </tbody>
  <!-- second model -->
  <tbody>
    <tr>
      ...
    </tr>
    <tr>
      ...
    </tr>
  </tbody>
</table>
```

In Marionette 2, I was able to use a `CompositeView`:

```js
const Item = Marionette.ItemView.extend({
  tagName: 'tbody',
  template: '#item'
});

const List = Marionette.CompositeView.extend({
  template: '#list',
  childViewContainer: 'table',
  childView: Item
});
```

With the following `#item` template:

```html
<tr>
  ...
</tr>
<tr>
  ...
</tr>
```

And the following `#list` template:

```html
<table>
  <thead>
    ...
  </thead>
  <!-- collection will be appended here -->
</table>
```

Because I used the `table` as the `childViewContainer`, the children were simply appended.

With `CompositeView` now being deprecated in Marionette 3, I'm not sure how to get the same end result without adding an extra layer. And I can't really add an extra layer if I want valid table HTML. From reading [the documentation](http://marionettejs.com/docs/master/marionette.collectionview.html#rendering-tables), it seems like the approved new way to do this is to use a `View` wrapper that shows a `CollectionView` in a `region` (probably using `replaceElement: true`). But unless I'm missing something, that still adds one layer (because the `CollectionView` needs an `el` of its own). The examples in the docs assume that the CollectionView can be a `tbody`, and the individual items can be `tr`s. But if I need the items to be `tbody`s, I need a way to render the collection without any wrapper of its own.

It seems like the concept of a `CollectionView` that both renders a template _and_ renders children in a `childViewContainer` has been removed in Marionette 3, without a good replacement. I suppose I could override the `CollectionView` `render` function to render a template as well as the children?

What am I missing here? Is there a recommended way to do table layout like this?

### Environment

1. Marionette version: 3.0.0
2. Backbone version: 1.3.3

**Update**: You can see the way I was doing this with Marionette v2--along with my attempt to upgrade to Marionette v3--here: https://jsfiddle.net/jgerigmeyer/L3ttz0k4/

**Comment by blikblum** [comment]
http://marionettejs.com/docs/master/marionette.collectionview.html#rendering-tables

**Comment by jgerigmeyer** [comment]
@blikblum Yes, I read that (and linked to it in my description). It doesn't address a case with no wrapper around the list of children.

**Comment by rafde** [comment]
@jgerigmeyer you brought up an interesting problem. Could you set up a fiddle for this, please? I'm thinking about using `document.createDocumentFragment()` for the el and something else that I am kicking around in my head.

**Comment by blikblum** [comment]
Sorry not reading your issue carefully

**Comment by jgerigmeyer** [comment]
@rafde I added a [jsfiddle](https://jsfiddle.net/jgerigmeyer/L3ttz0k4/1/) to the description.

**Comment by paulfalgout** [comment]
Use `CollectionView#addChildView`
https://jsfiddle.net/L3ttz0k4/3/

**Comment by jgerigmeyer** [comment]
@paulfalgout Won't that cause indexing bugs with a sorted collection if models are added later? It also adds complexities with sorting/reordering, changing the `filter`, etc.

https://jsfiddle.net/jgerigmeyer/L3ttz0k4/5/

**Comment by rafde** [comment]
this issue is looking related to https://github.com/marionettejs/backbone.marionette/issues/3128 where
```
<table>
  <thead>
    <!-- don't remove me, bro -->
  </thead>
  <!-- collection will be appended here -->
</table>
```

anyhow https://jsfiddle.net/yj2engmc/1/

**Comment by paulfalgout** [comment]
@jgerigmeyer for this fiddle you can use `onRenderChildren` https://jsfiddle.net/L3ttz0k4/6/

but yes.. you will have to manage the view added in regards to sorting / filtering

But in this case collectionView sorting/filtering will work without any effort.

you should be able to add/remove to the collection with sorting..

also
```js
list3.viewComparator = 'thing2';
list3.resortView();
```
will work as well as 

```js
list3.setFilter(function(child){ return child.get('thing1') !== 'Baz'; });
```

If you ever get a more complicated addition you might need to handle the index of the added view inside the filter.. or _possibly_ do some more complicated switcheroos for maintaining a sort order (like say you wanted insert headings every 50 rows) 

but in those cases I don't think CompositeView is any better anyhow.

**Comment by paulfalgout** [comment]
Just dropping a brief code comparison for visuals.  I think this supports what is being asked:

In v3
```js
const List = Mn.CompositeView.extend({
  childView: Item,
  onRenderChildren () {
  	this.addChildView(myHeadView, 0);
  }
});
```

In v2:
```js
const List = Mn.CompositeView.extend({
  template: '#list',
  childViewContainer: 'table',
  childView: Item
});
```

**Comment by paulfalgout** [comment]
@jgerigmeyer is this resolved?

**Comment by jgerigmeyer** [comment]
@paulfalgout Ok, yeah I think that works. If I want to use `reorderOnSort: true`, I also need to fire the `addChildView` in an `onReorder` handler, but that's not a big deal.

Generally, I find @rafde's solution a bit cleaner, because I don't need to worry about introducing indexing bugs down the road, and I'm not using the undocumented `addChildView` fn.

Thanks all! I'll leave it up to the maintainers to decide whether this use-case deserves better documentation or a cleaner API.

**Comment by paulfalgout** [comment]
`addChildView` is documented here: http://marionettejs.com/docs/master/marionette.collectionviewadvanced.html#collectionviews-addchildview

Whichever solution works for you as long as we're covering the edge cases for remove `CompositeView`.  But note that attaching the collection view's el after the render will mean lifecycle events of the child views won't be running.  the child views will not receive onAttach events nor with they be aware that they are attached.

**Comment by jgerigmeyer** [comment]
For anyone following along at home, I want to note that the recommended solution above (https://github.com/marionettejs/backbone.marionette/issues/3150#issuecomment-243997697) isn't quite as straightforward as it seems.

1. If you want `addChildView` to work consistently with sorting/filtering of the collection, you need to call it both `onRender` and `onReorder`.
2. Because of view indexing bugs in Marionette (specifically item 2 in https://github.com/marionettejs/backbone.marionette/issues/2948#issue-142114832), calling `addChildView(myView, 0)` will not consistently prepend `myView` to the beginning of the list of children. Instead, you need to find the childview with the lowest `_index`, and use that instead:

```js
const minIndex = _.min(this.children.map((child) => child._index));
this.addChildView(myView, minIndex);
```

**Comment by paulfalgout** [comment]
True for 1 if using reorderOnSort. 

I think for 2, that might be fixed already in the upcoming 3.1 and if not I would expect it in 3.2. @rafde is the index issue already resolved or is it an add issue?  Also I have no idea what will happen off hand, but you could try `this.addChildView(myView,  -Infinity)` in the meantime... Might be hacky to get around the real issue, but cheap and consistent.

**Comment by rafde** [comment]
😢 I can't resolve re-index until I properly resolve https://github.com/marionettejs/backbone.marionette/pull/3173

I can prioritize that work instead of the benchmark stuff, seems to be a need to resolve that PR.
